app/src/app-state-reducer.js

```diff
@@ -0,0 +1,33 @@
+import BN from 'bn.js'
+
+function transformTokenData(tokenData) {
+  const { decimals, totalSupply } = tokenData
+  return {
+    ...tokenData,
+    decimals: new BN(decimals),
+    numDecimals: parseInt(decimals, 10),
+    tokenDecimalsBase: new BN(10).pow(new BN(decimals)),
+    totalSupply: new BN(totalSupply),
+  }
+}
+
+function appStateReducer(state) {
+  if (state === null) {
+    return { syncing: true }
+  }
+
+  const { holders, outsideToken, wrappedToken } = state
+
+  return {
+    ...state,
+    holders:
+      holders &&
+      holders
+        .map(holder => ({ ...holder, balance: new BN(holder.balance) }))
+        .sort((a, b) => b.balance.cmp(a.balance)),
+    outsideToken: outsideToken && transformTokenData(outsideToken),
+    wrappedToken: wrappedToken && transformTokenData(wrappedToken),
+  }
+}
+
+export default appStateReducer
```

Comment:
The main change with this reducer is to map numerical values obtained from the background script into easy-to-use numbers in the frontend.

We generally pass numbers from the background script to the frontend as strings, since they can be very large numbers (`decimals` are safe because they must be very small numbers).

In the reducer, we then migrate these strings into BN instances, so we can easily apply math and other arithmetic operations if we have to.